app/(chat)/actions.tsx

```diff
@@ -0,0 +1,169 @@
+'use server'
+
+import { kv } from '@vercel/kv'
+import {
+  OpenAIStream,
+  experimental_StreamingReactResponse,
+  experimental_StreamData,
+  type Message
+} from 'ai'
+import OpenAI from 'openai'
+
+import { auth } from '@/auth'
+import { nanoid } from '@/lib/utils'
+import {
+  ChatCompletionCreateParams,
+  ChatCompletionMessageParam
+} from 'openai/resources'
+
+import { DataView } from '@/components/data-view'
+
+const openai = new OpenAI({
+  apiKey: process.env.OPENAI_API_KEY
+})
+
+const functions: ChatCompletionCreateParams.Function[] = [
+  {
+    name: 'get_current_weather',
+    description: 'Get the current weather',
+    parameters: {
+      type: 'object',
+      properties: {
+        location: {
+          type: 'string',
+          description: 'The city and state, e.g. San Francisco, CA'
+        },
+        format: {
+          type: 'string',
+          enum: ['celsius', 'fahrenheit'],
+          description:
+            'The temperature unit to use. Infer this from the users location.'
+        }
+      },
+      required: ['location', 'format']
+    }
+  },
+  {
+    name: 'create_image',
+    description: 'Create an image for the given description',
+    parameters: {
+      type: 'object',
+      properties: {
+        description: {
+          type: 'string',
+          description: 'Description of what the image should be.'
+        }
+      },
+      required: ['description']
+    }
+  }
+]
+
+export async function handleChat(
+  meta: {
+    id: string
+    previewToken?: string
+  },
+  chat: {
+    messages: Message[]
+  }
+) {
+  const userId = (await auth())?.user.id
+
+  if (!userId) {
+    return new Response('Unauthorized', {
+      status: 401
+    })
+  }
+
+  if (meta.previewToken) {
+    openai.apiKey = meta.previewToken
+  }
+
+  const data = new experimental_StreamData()
+  const response = await openai.chat.completions.create({
+    model: 'gpt-3.5-turbo',
+    messages: chat.messages.map(m => ({
+      content: m.content,
+      role: m.role
+    })) as ChatCompletionMessageParam[],
+    functions,
+    stream: true
+  })
+
+  const stream = OpenAIStream(response, {
+    onFinal() {
+      data.close()
+    },
+    async experimental_onFunctionCall({ name, arguments: args }) {
```

Comment:
I would consider updating this to usexperimental_onToolCall and possibly consider moving kv into onFinal to avoid unnecessary updates.